File: ui/src/app/analysis/pages/map/_page/map-move.ts
import {SelectedFeatureHolder} from "./selected-feature-holder";
import Map from 'ol/Map';
import {MapState} from "./map-state";
import Select from 'ol/interaction/Select';
import {pointerMove} from 'ol/events/condition';
import Style from 'ol/style/Style';

export class MapMove {

  readonly interaction = new Select({
    condition: pointerMove,
    multi: false,
    style: new Style() // this overrides the normal openlayers default edit style
  });

  constructor(
    map: Map,
    state: MapState,
    selectionHolder: SelectedFeatureHolder,
    refreshCallback: () => void) {

    this.interaction.on("select", (e) => this.selectListener(map, state, selectionHolder, refreshCallback, e));
  }

  private selectListener(
    map: Map,
    mapState: MapState,
    selectionHolder: SelectedFeatureHolder,
    refreshCallback: () => void,
    e /*: ol.interaction.select.Event*/): boolean {

    if (e.selected.length > 0) {
      map.getTargetElement().setAttribute("style", "cursor: pointer")
    }
    else {
      map.getTargetElement().setAttribute("style", "cursor: default")
    }

    for (let feature of e.deselected) {
      const layer = feature.get("layer");
      if (layer.endsWith("route")) {
        mapState.highlightedRouteId = null
      }
      else if (layer.endsWith("node")) {
        mapState.highlightedNodeId = null
      }
    }

    for(let feature of e.selected) {
      const layer = feature.get("layer");
      const id = feature.get("id");
      if (layer.endsWith("route")) {
        const id2 = id.substring(0, id.indexOf('-'));
        mapState.highlightedRouteId = id2;
      }
      else if (layer.endsWith("node")) {
        mapState.highlightedRouteId = null;
        mapState.highlightedNodeId = id;
      }
    }

    refreshCallback();
    return true;
  }

}

File: production/client/src/main/scala/kpn/client/common/map/vector/MapMove.scala
package kpn.client.common.map.vector

class MapMove(state: MapState, map: ol.Map, refreshCallback: () => Unit) {

  val interaction: ol.interaction.Select = {
    val select = new ol.interaction.Select(
      olx.interaction.SelectOptions(
        condition = ol.events.condition.Condition.pointerMove _,
        multi = false,
        style = ol.style.Style() // this overrides the normal openlayers default edit style
      )
    )
    select.on("select", scalaPointerMoveListener _)
    select
  }

  private def scalaPointerMoveListener(e: ol.interaction.select.Event): Boolean = {

    if (e.selected.length > 0) {
      map.getTargetElement().setAttribute("style", "cursor: pointer")
    }
    else {
      map.getTargetElement().setAttribute("style", "cursor: default")
    }

    e.deselected.foreach { feature: ol.Feature =>
      val layer = feature.asInstanceOf[ol.render.Feature].get("layer").toString
      if (layer.endsWith("route")) {
        state.highlightedRouteId = None
      }
      else if (layer.endsWith("node")) {
        state.highlightedNodeId = None
      }
    }

    e.selected.foreach { feature: ol.Feature =>
      val layer = feature.asInstanceOf[ol.render.Feature].get("layer").toString
      if (layer.endsWith("route")) {
        val id = feature.asInstanceOf[ol.render.Feature].get("id").toString
        val id2 = id.takeWhile(_ != '-') + "-"
        state.highlightedRouteId = Some(id2)
      }
      else if (layer.endsWith("node")) {
        state.highlightedRouteId = None
        state.highlightedNodeId = Some(feature.asInstanceOf[ol.render.Feature].get("id").toString)
      }
    }

    refreshCallback()
    true
  }

}
